plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs

```diff
@@ -71,11 +71,11 @@ class (Typeable uni, Typeable fun, Bounded fun, Enum fun, Ix fun) => ToBuiltinMe
     type CostingPart uni fun
 
     -- | Get the 'BuiltinMeaning' of a built-in function.
-    toBuiltinMeaning :: HasConstantIn uni val => fun -> BuiltinMeaning val (CostingPart uni fun)
+    toBuiltinMeaning :: HasConstantIn uni val => Version ann -> fun -> BuiltinMeaning val (CostingPart uni fun)
```

Comment:
I think we should not use the `Version` type for this. It's not 100% clear what it means, but I think it should refer to the version of plutus core the _language_, which is naturally independent of the set of builtins. Yes, this is hopelessly confused by the fact that that ledger also talks about plutus language versions, but tough.

Maybe a new type in `Default/Builtins.hs`, `BuiltinsVersion` or something? Because that's actually the thing we're changing. And then in the V2 API we would say we're using the V2 builtins.

The annoying thing is that we need a type here in this typeclass method. I guess we could stick an associated type in there, so we have something like:
```
class ... ToBuiltinMeaning uni fun where
   type BuiltinVersion uni fun
   toBuiltinMeaning :: HasConstantIN uni val => BuiltinVersion fun -> BuiltinMeaning val (CostingPart uni fun)
```

Reply:
Yeah, although I thought you were mostly talking about pushing things into the types. I mostly just wanted us to not reuse `Version`, I don't have so much opinion about whether we should do it at the value or the type level.

Reply:
Or we could simply have 

```haskell
data Versioned v a = Versioned !v !a

data DefaultFunVersion
    = DefaultFunV1
    | DefaultFunV2

instance uni ~ DefaultUni => ToBuiltinMeaning uni (Versioned DefaultFunVersion DefaultFun) where
    type CostingPart uni (Versioned DefaultFunVersion DefaultFun) = BuiltinCostModel
```

Reply:
Also doesn't force us to explicitly ignore a redundant argument in every builtin other than `ConsByteString`. Doesn't impose any versioning on sets of builtins where we don't care about versioning, like the `NopFun` and `ExtensionFun` sets. And may also fix the performance problem right away.

Reply:
> Right, but then we'd have to actually put those `Versioned` constructors in the program everywhere, right?

No, you can always recover `toBuiltinMeaning :: DefaultFunVersion -> DefaultFun -> blah-blah-blah` from the `Versioned`-based interface by simply packaging the version with the builtin using the `Versioned` constructor.

Reply:
I.e.  `f :: DefaultFunVersion -> DefaultFun -> r` is the same as `f :: Versioned DefaultFunVersion DefaultFun -> r`, except the latter is more convenient as it doesn't require changing as much machinery and imposing versioning on everything.